### PR TITLE
Errors with cart service

### PR DESCRIPTION
**Pull request by nilmadhab**

### CODE REVIEW

File: src/main/java/com/webtutsplus/ecommerce/controller/CartController.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    @Autowired
+    private ProductService productService;
+
+    @Autowired
+    private AuthenticationService authenticationService;
+
+    @PostMapping("/add")
+    public ResponseEntity<ApiResponse> addToCart(@RequestBody CartDto cartDto, @RequestParam("token") String token) {
+        int userId = authenticationService.getUser(token).getId();
+        System.out.println("DTO : " + cartDto.toString());
```
Comment (nilmadhab) [inline src/main/java/com/webtutsplus/ecommerce/controller/CartController.java:38]: remove the logs

File: src/main/java/com/webtutsplus/ecommerce/controller/CartController.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            return new ResponseEntity<ApiResponse>(new ApiResponse(false, "Item is invalid"), HttpStatus.CONFLICT);
+        }
+        Product product = optionalProduct.get();
+        cartService.addToCart(cartDto,product,userId);
+
+        return new ResponseEntity<ApiResponse>(new ApiResponse(true, "Added to cart"), HttpStatus.CREATED);
+
+    }
+    @GetMapping("/")
+    public ResponseEntity<List<CartDto>> getCartItems() {
```
Comment (nilmadhab) [inline src/main/java/com/webtutsplus/ecommerce/controller/CartController.java:51]: user field is missing

File: src/main/java/com/webtutsplus/ecommerce/controller/CartController.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        Optional<Product> optionalProduct = productService.readProduct(cartDto.getProductId());
+        if (!optionalProduct.isPresent()) {
+            return new ResponseEntity<ApiResponse>(new ApiResponse(false, "product is invalid"), HttpStatus.CONFLICT);
+        }
+        Product product = optionalProduct.get();
+        cartService.updateCartItem(itemID, cartDto, product);
+        return new ResponseEntity<ApiResponse>(new ApiResponse(true, "Product has been updated"), HttpStatus.OK);
+    }
+
+    @DeleteMapping("/delete/{cartItemId}")
```
Comment (nilmadhab) [inline src/main/java/com/webtutsplus/ecommerce/controller/CartController.java:67]: user id is missing

File: src/main/java/com/webtutsplus/ecommerce/repository/CartRepository.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+import com.webtutsplus.ecommerce.model.Cart;
+import com.webtutsplus.ecommerce.model.WishList;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+@Repository
+public interface CartRepository extends JpaRepository<Cart, Integer> {
+
+    List<Cart> findAllByUserIdOrderByCreatedDateDesc(Integer userId);
```
Comment (nilmadhab) [inline src/main/java/com/webtutsplus/ecommerce/repository/CartRepository.java:12]: cart should be for a particular user

File: src/main/java/com/webtutsplus/ecommerce/service/CartService.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        return cart;
+    }
+
+    public void addToCart(CartDto cartDto, Product product,int userId) {
+        Cart cart = getCartFromDto(cartDto, product,userId);
+        cartRepository.save(cart);
+    }
+
+    public List<CartDto> listCartItems() {
+        List<Cart> cartList = cartRepository.findAll();
```
Comment (nilmadhab) [inline src/main/java/com/webtutsplus/ecommerce/service/CartService.java:44]: should be for a particular user

File: src/main/resources/application.properties [hunk 6]
```diff
@@ -1,6 +1,6 @@
 spring.datasource.url=jdbc:mysql://localhost:3306/ecommerce
 spring.datasource.username=root
-spring.datasource.password=root
+spring.datasource.password=
```
Comment (nilmadhab) [inline src/main/resources/application.properties:3]: remove this chnage

File: src/main/java/com/webtutsplus/ecommerce/service/ProductService.java [hunk 7]
```diff
@@ -45,4 +47,12 @@ public void updateProduct(long productID, ProductDto productDto, Category catego
         product.setId(productID);
         productRepository.save(product);
     }
+
+
+    public Optional<Product> readProduct(Long productId) {
```
Comment (nilmadhab) [inline src/main/java/com/webtutsplus/ecommerce/service/ProductService.java:52]: rename to getProductById

File: src/main/java/com/webtutsplus/ecommerce/service/CartService.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    public CartCost listCartItems(int user_id) {
+        List<Cart> cartList = cartRepository.findAllByUserIdOrderByCreatedDateDesc(user_id);
+        List<CartDto> cartItems = new ArrayList<>();
+        for (Cart cart:cartList){
+            CartDto cartDto = getDtoFromCart(cart);
+            cartItems.add(cartDto);
+        }
+        int totalCost = 0;
```
Comment (nilmadhab) [inline src/main/java/com/webtutsplus/ecommerce/service/CartService.java:43]: should be double, as price can be double

File: src/main/java/com/webtutsplus/ecommerce/controller/CartController.java [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+        Product product = optionalProduct.get();
+        int userId = authenticationService.getUser(token).getId();
+//        cartService.updateCartItem(itemID,userId,quantity);
+        cartService.updateCartItem(itemID, cartDto,product,userId,quantity);
+        return new ResponseEntity<ApiResponse>(new ApiResponse(true, "Product has been updated"), HttpStatus.OK);
+    }
+
+    @DeleteMapping("/delete/{cartItemId}")
+    public ResponseEntity<ApiResponse> deleteCartItem(@PathVariable("cartItemId") int itemID,@RequestParam("token") String token){
+        int userId = authenticationService.getUser(token).getId();
```
Comment (nilmadhab) [inline src/main/java/com/webtutsplus/ecommerce/controller/CartController.java:68]: chek for valid user, throw exception if token is not valid

File: src/main/java/com/webtutsplus/ecommerce/controller/CartController.java [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+
+    @Autowired
+    private ProductService productService;
+
+    @Autowired
+    private AuthenticationService authenticationService;
+
+    @PostMapping("/add")
+    public ResponseEntity<ApiResponse> addToCart(@RequestBody CartDto cartDto, @RequestParam("token") String token) {
+        int userId = authenticationService.getUser(token).getId();
```
Comment (nilmadhab) [inline src/main/java/com/webtutsplus/ecommerce/controller/CartController.java:35]: check for valid user, throw an exception if token is not valid

File: src/main/java/com/webtutsplus/ecommerce/service/CartService.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+        cart.setCreatedDate(new Date());
+        cartRepository.save(cart);
+    }
+    public static Cart getCartFromDto(CartDto cartDto, Product product) {
+        Cart cart = new Cart(cartDto, product);
+        return cart;
+    }
+
+    public void deleteCartItem(int id,int userId){
+        cartRepository.deleteById(id);
```
Comment (nilmadhab) [inline src/main/java/com/webtutsplus/ecommerce/service/CartService.java:71]: handle invalid id case

File: src/main/java/com/webtutsplus/ecommerce/service/ProductService.java (hunk 7 above)
Comment (SudhanshuGulhane) [reply src/main/java/com/webtutsplus/ecommerce/service/ProductService.java:52]: solved

File: src/main/java/com/webtutsplus/ecommerce/service/CartService.java (hunk 8 above)
Comment (SudhanshuGulhane) [reply src/main/java/com/webtutsplus/ecommerce/service/CartService.java:43]: solved

File: src/main/java/com/webtutsplus/ecommerce/controller/CartController.java (hunk 9 above)
Comment (SudhanshuGulhane) [reply src/main/java/com/webtutsplus/ecommerce/controller/CartController.java:68]: added try catch

File: src/main/java/com/webtutsplus/ecommerce/controller/CartController.java (hunk 10 above)
Comment (SudhanshuGulhane) [reply src/main/java/com/webtutsplus/ecommerce/controller/CartController.java:35]: added try catch

File: src/main/java/com/webtutsplus/ecommerce/controller/CartController.java (hunk 9 above)
Comment (SudhanshuGulhane) [reply src/main/java/com/webtutsplus/ecommerce/controller/CartController.java:68]: added exception part

File: src/main/java/com/webtutsplus/ecommerce/controller/CartController.java (hunk 10 above)
Comment (SudhanshuGulhane) [reply src/main/java/com/webtutsplus/ecommerce/controller/CartController.java:35]: here also added the exception part

Comment (SudhanshuGulhane) [reply src/main/java/com/webtutsplus/ecommerce/controller/CartController.java:35]: will commit the changes

File: src/main/java/com/webtutsplus/ecommerce/service/CartService.java (hunk 11 above)
Comment (SudhanshuGulhane) [reply src/main/java/com/webtutsplus/ecommerce/service/CartService.java:71]: handled the invalid id case..

**Pull request merged**